imbi/endpoints/integrations/google.py

```diff
@@ -0,0 +1,143 @@
+import logging
+import re
+import urllib.parse
+
+import google.auth.exceptions
+import sprockets.mixins.http
+import yarl
+from google.oauth2 import id_token
+from google.auth.transport import requests
+
+import imbi.user
+from imbi import errors, oauth2
+from imbi.endpoints import base
+
+LOGGER = logging.getLogger(__name__)
+
+
+class RedirectHandler(sprockets.mixins.http.HTTPClientMixin,
+                      base.RequestHandler):
+    integration: oauth2.OAuth2Integration
+
+    NAME = 'google-redirect'
+
+    GET_TOKEN_SQL = re.sub(r'\s+', ' ', """
+        SELECT * FROM v1.user_oauth2_tokens
+         WHERE integration = 'google'
+           AND external_id = %(external_id)s
+    """)
+
+    ADD_TOKEN_SQL = re.sub(r'\s+', ' ', """
+        INSERT INTO v1.user_oauth2_tokens
+                    (username, integration, external_id,
+                     access_token, refresh_token)
+             VALUES (%(username)s, 'google', %(external_id)s,
+                     %(access_token)s, %(refresh_token)s)
+        ON CONFLICT (integration, external_id)
+                 DO UPDATE SET access_token = EXCLUDED.access_token,
+                              refresh_token = EXCLUDED.refresh_token
+    """)
+
+    UPDATE_TOKEN_SQL = re.sub(r'\s+', ' ', """
+       UPDATE v1.user_oauth2_tokens
+          SET access_token = %(access_token)s,
+              username = %(username)s
+        WHERE integration = 'google'
+          AND external_id = %(external_id)s
+    """)
+
+    async def prepare(self) -> None:
+        await super().prepare()
+        if not self._finished:
+            self.integration = await oauth2.OAuth2Integration.by_name(
+                self.application, 'google')
+            if not self.integration:
+                raise errors.IntegrationNotFound('google')
+
+    async def get(self):
+        code = self.get_query_argument('code', default=None)
```

Comment:
I feel like we should be passing something verifiable in the `state` parameter but I'm not completely sure if it would buy us anything in this case.  The usual flow for OAuth 2 stuff is:

1. stash stuff into something like redis about the user that we are expecting using a UUID as the key
2. pass the key as the `state` parameter in the redirect to the authorization endpoint
3. pull the `state` parameter out of the URL in the redirect handler
4. look up the state in redis, verify it, and delete it
5. redeem the authorization code for an access/refresh token pair
6. stash them somewhere
7. redirect back into the UI

We are in the weird place where we start the authorization flow in the UI code and handle token redemption in the API code.  There isn't anything that ensures that we are redeeming a token for the UI as opposed to some other chunk of code somewhere.  I'm probably overthinking this one since we are not returning tokens outside of the session cookie.  Though that does grant access to the API 🤔
```
curl -H 'Cookie: Path=/;Domain=imbi.aweber.io;session=2|....' https://imbi.aweber.io/projects
```

Meh ... it is restricted to the `imbi.aweber.io` domain so we are pretty safe here.  We might want to make it `HttpOnly` if we can since none of the JS code should be looking at it.

Reply:
We should certainly set `HttpOnly`, good catch.

Perhaps I can refactor the flow to start in the backend. Instead of going directly to the Google auth URL, we'll navigate to `/ui/login/google`, which on the backend will generate the `state` token and then redirect.